crates/gpui/src/platform/windows/direct_write.rs

```diff
@@ -0,0 +1,1286 @@
+                text_layout
+                    .SetFontStyle(font_info.font_face.GetStyle(), text_range)
+                    .unwrap();
+                text_layout
+                    .SetFontWeight(font_info.font_face.GetWeight(), text_range)
+                    .unwrap();
+                text_layout
+                    .SetTypography(&font_info.features, text_range)
+                    .unwrap();
+            }
+
+            let mut runs = Vec::new();
+            let renderer_context = RendererContext {
+                text_system: self,
+                index_converter: StringIndexConverter::new(text),
+                runs: &mut runs,
+                utf16_index: 0,
+                width: 0.0,
+            };
+            text_layout
+                .Draw(
+                    Some(&renderer_context as *const _ as _),
+                    &text_renderer.0,
+                    0.0,
+                    0.0,
+                )
+                .unwrap();
+            let width = px(renderer_context.width);
+
+            LineLayout {
+                font_size,
+                width,
+                ascent,
+                descent,
+                runs,
+                len: text.len(),
+            }
+        }
+    }
+
+    fn font_metrics(&self, font_id: FontId) -> FontMetrics {
+        unsafe {
+            let font_info = &self.fonts[font_id.0];
+            let mut metrics = std::mem::zeroed();
+            font_info.font_face.GetMetrics2(&mut metrics);
+
+            FontMetrics {
+                units_per_em: metrics.Base.designUnitsPerEm as _,
+                ascent: metrics.Base.ascent as _,
+                descent: -(metrics.Base.descent as f32),
+                line_gap: metrics.Base.lineGap as _,
+                underline_position: metrics.Base.underlinePosition as _,
+                underline_thickness: metrics.Base.underlineThickness as _,
+                cap_height: metrics.Base.capHeight as _,
+                x_height: metrics.Base.xHeight as _,
+                bounding_box: Bounds {
+                    origin: Point {
+                        x: metrics.glyphBoxLeft as _,
+                        y: metrics.glyphBoxBottom as _,
+                    },
+                    size: Size {
+                        width: (metrics.glyphBoxRight - metrics.glyphBoxLeft) as _,
+                        height: (metrics.glyphBoxTop - metrics.glyphBoxBottom) as _,
+                    },
+                },
+            }
+        }
+    }
+
+    unsafe fn get_glyphrun_analysis(
+        &self,
+        params: &RenderGlyphParams,
+    ) -> windows::core::Result<IDWriteGlyphRunAnalysis> {
+        let font = &self.fonts[params.font_id.0];
+        let glyph_id = [params.glyph_id.0 as u16];
+        let advance = [0.0f32];
+        let offset = [DWRITE_GLYPH_OFFSET::default()];
+        let glyph_run = DWRITE_GLYPH_RUN {
+            fontFace: unsafe { std::mem::transmute_copy(&font.font_face) },
```

Comment:
Rather than transmute, can we construct these in an ordinary way? e.g. `ManuallyDrop::new(Some(font.font_face.clone()))`

Reply:
`transmute_copy` is actually the "correct" way, `ManuallyDrop::new(Some(font.font_face.clone()))` will lead to a memory leak, the `font_face` will never be released.

The maintainer of `windows-rs` said:
> That would be the strict equivalence of what you would do in C++ and pass in a raw interface pointer.

Please refer to the complete explanation here: https://github.com/microsoft/windows-rs/issues/2386